### ISSUE TITLE
The workflow should be updated after the uptreaming finished

### ISSUE THREAD

**Issue by lazyparser** [riscv-collab/v8#442]
Now the RV64 porting had been merged into upstream. We should consider the workflow after the upstream. There is some brainstorming: 

1. the wiki is still useful and should be continuously updated. some contents (testing status, etc) would be modified/removed.
2. Small patches could be submitted to chromium/gerrit directly.
3. The CI for v8-riscv project is still useful, we may want to keep it.
4. Big features like RVV, I have no idea whether should go upstream directly or put the WIP patch set on v8-riscv/v8.
5. NodeJS is our next target in roadmap 2021. We should start to add more build/test infrastructure and effort on it.
6. Current open pull requests ? go upstream directly? Rebasing v8 upstream is definitely laborious. Should try to void it.

**Comment by fw-brice** [comment]
Thanks for starting this discussion Wei. I think it would be a good idea to work through the larger changes in this repo, then once it is ready, rebase and open an upstream review. 

We also need to understand the process that we need to follow for the upstream reviews. I will send an email to the Google team to help point us to the documentation for that.

**Comment by fw-brice** [comment]
I added my initial thoughts to the wiki - https://github.com/v8-riscv/v8/wiki/Upstream-Workflow. Let's discuss here and finalize this plan.

**Comment by lazyparser** [comment]
> I added my initial thoughts to the wiki - https://github.com/v8-riscv/v8/wiki/Upstream-Workflow. Let's discuss here and finalize this plan.

Thank Brice for writing these down. I have an another opinion: Instead of using the `riscv64` as the unique developer branch, we may better to use one `feature branch` for exactly on `feature issue` or `bug issue`.

For the long term feature, say rvv, we can create a `rvv-dev`, and merge all WIP patches/branches into the branch, e.g. `rvv-instr-123` merged into `rvv-dev` after CI and review. And eventually the `rvv-dev` branch go upstream for review and merge. the `riscv64` branch could be sticking to  `the last good version for RISC-V on master/main branch of upstream`.

**Comment by lazyparser** [comment]
and for the support for nodejs and other downstream projects, we may also need to considering the backporting: e.g. if a LTS version of nodejs uses V8 `v8.9` and other LTS uses V8 `v9.2`, then we may need to maintain `riscv64-v8.9-branch` and `riscv64-v9.2-branch` for the `Last Good Commit tracking`.
(The backporting is not an urgent thing to deal with.)

**Comment by luyahan** [comment]
Should we open a same issue about important problem on chromium issues？

**Comment by fw-brice** [comment]
> For the long term feature, say rvv, we can create a rvv-dev, and merge all WIP patches/branches into the branch, e.g. rvv-instr-123 merged into rvv-dev after CI and review. And eventually the rvv-dev branch go upstream for review and merge. the riscv64 branch could be sticking to the last good version for RISC-V on master/main branch of upstream.

I think that is essentially the same as what I was suggesting, with the addition of sub-branches on the feature branches. We do work on feature branches, then once ready, those branches get pushed upstream for review. Master is always in sync with upstream, and riscv64 is always the latest commit from master which passes all RISC-V tests.

**Comment by lazyparser** [comment]
> > For the long term feature, say rvv, we can create a rvv-dev, and merge all WIP patches/branches into the branch, e.g. rvv-instr-123 merged into rvv-dev after CI and review. And eventually the rvv-dev branch go upstream for review and merge. the riscv64 branch could be sticking to the last good version for RISC-V on master/main branch of upstream.
> 
> I think that is essentially the same as what I was suggesting, with the addition of sub-branches on the feature branches. We do work on feature branches, then once ready, those branches get pushed upstream for review. Master is always in sync with upstream, and riscv64 is always the latest commit from master which passes all RISC-V tests.

Yes. It would be helpful if we can emphasis on `do not merge the PR here, go chromium/gerrit`.
LGTM. Thanks.

**Comment by qjivy** [comment]
LGTM

**Comment by derekztu22** [comment]
> > For the long term feature, say rvv, we can create a rvv-dev, and merge all WIP patches/branches into the branch, e.g. rvv-instr-123 merged into rvv-dev after CI and review. And eventually the rvv-dev branch go upstream for review and merge. the riscv64 branch could be sticking to the last good version for RISC-V on master/main branch of upstream.
> 
> I think that is essentially the same as what I was suggesting, with the addition of sub-branches on the feature branches. We do work on feature branches, then once ready, those branches get pushed upstream for review. Master is always in sync with upstream, and riscv64 is always the latest commit from master which passes all RISC-V tests.

Are these the same? I will explain my understanding from reading Brices document and what I can understand from Wei's proposal and please let me know if my understanding is incorrect.

From what I understand from Brice's [document](https://github.com/v8-riscv/v8/wiki/Upstream-Workflow) this is the proposed git structure (forgive me for the ugly diagrams):

```
                        v8/master
                             |             
                         v8-riscv/master   
                             |
                      v8-riscv/riscv64                     
                    /            \
 v8-riscv/new-feature0  v8-riscv/new-feature1
      /               \
v8-riscv/f0-sub0      v8-riscv/f0-sub1
```

where:
v8-riscv/master is always synced but could be broken.
v8-riscv/riscv64 is synced to the latest commit that passes RISC-V tests from v8-riscv/master

whereas Wei's proposed git structure is this

```
                             v8/master
                           /          \             
                  v8-riscv/riscv64   v8-riscv/master   
                                    /            \
                 v8-riscv/new-feature0  v8-riscv/new-feature1
                /               \
 v8-riscv/f0-sub0      v8-riscv/f0-sub1  
```                

where: 
v8-riscv/master is always synced but could be broken.
v8-riscv/riscv64 is synced to v8/master so is always a working "master" branch.

**Comment by fw-brice** [comment]
Yes, that looks right for my proposal, and it clarifies how a branch that fixes an upstream issue in master would work. Thanks Derek.

**Comment by derekztu22** [comment]
Then aren't the two proposals different in terms of when a broken patch is pushed to v8/master?

For Brice's proposal when a broken patch is pushed we will still have a branch that can contain all the newest features and can continue gaining newer features while we fix the broken patch; in this case the branch would be v8-riscv/riscv64.

For Wei's proposal when we get a broken patch I assume we would stop pushing while we wait for a fix to v8-riscv/master.

It's very nitpicky and honestly probably not that big an issue, but I think the discussion here about the differences could help us write better documentation when we do choose one of the two approaches.

**Comment by fw-brice** [comment]
Wei's suggestion says "the riscv64 branch could be sticking to the last good version for RISC-V on master/main branch of upstream." Isn't that saying the same thing as mine?

**Comment by derekztu22** [comment]
From what I understand from your suggestion is that we will be able to continue updating the v8-riscv/riscv64 branch with new features. While with Wei's suggestion all v8-riscv/riscv64 is doing is tracking the last working version of RISC-V from v8/master.

**Comment by fw-brice** [comment]
Oh, no actually, I was also suggesting that feature branches are based off of riscv64, but they are not merged back into riscv64, only merged upstream, once riscv64 is in sync with master.

master is only updated when upstream is updated. When master is updated, riscv64 is updated to the latest commit that passes all RISC-V tests. So if developers are working on a feature while master is broken, they will base their feature branch off of riscv64, but then once the issue is fixed on master, riscv64 is updated and all feature branches must be rebased. 

The advantage of this is that it keeps all of our features separated, so that the upstream reviews are cleaner, with just one change at a time.

**Comment by derekztu22** [comment]
Okay I got it, thanks for the clarification!

**Comment by fw-brice** [comment]
I'll take another pass at the wiki page to see if I can clarify that better.

**Comment by lazyparser** [comment]
Hi all

I think we have made a consensus already. I'll start to work on the automation sync tools.

Can we close the issue now?

**Comment by fw-brice** [comment]
That sounds good to me. Thanks Wei. Feel free to make changes in my CI tool, or let me know what changes you'd like me to make there.

**Comment by lazyparser** [comment]
All developments are moved to upstream.
